File: web/src/lib/scoring/features.ts
// Online feature engineering — the exact counterpart of sentinel_ml/features.py
// for a single incoming event. The offline pipeline streams the whole log; here
// we reconstruct the same per-user / per-IP state from rows the feature store
// (Postgres) returns, under the identical causal constraint: only information
// available *before* this event may be used.
//
// Windowing conventions kept identical to Python:
//   * failed_attempts_* and logins_user_1h count PRIOR events only.
//   * distinct_*_24h count prior-in-window values PLUS the current event.
//   * *_is_known compares against everything seen strictly before this event.

import { haversineKm } from "./geo";
import { clip, encodeCategorical, type FeatureSpec } from "./spec";
import type { HistoryRow, RawEvent } from "./types";

const HOUR_MS = 60 * 60 * 1000;
const DAY_MS = 24 * HOUR_MS;

function isOffHours(hour: number, start: number, end: number): number {
  const endMod = end % 24;
  if (start === endMod) return 0;
  const active = start < endMod ? hour >= start && hour < endMod : hour >= start || hour < endMod;
  return active ? 0 : 1;
}

export interface ComputedFeatures {
  vector: number[];
  named: Record<string, number>;
}

/**
 * @param event        the event being scored
 * @param userHistory  this user's prior events (ascending ts, all < event.ts)
 * @param ipFailLast1h  prior failed attempts on this IP within the last hour (any user)
 */
export function computeFeatures(
  spec: FeatureSpec,
  event: RawEvent,
  userHistory: HistoryRow[],
  ipFailLast1h: number,
): ComputedFeatures {
  const now = Date.parse(event.ts);
  const jsDay = new Date(now).getUTCDay(); // Sun=0
  const dow = (jsDay + 6) % 7; // pandas: Mon=0
  const hour = new Date(now).getUTCHours();

  const prior = userHistory.filter((h) => Date.parse(h.ts) < now);
  const last = prior.length ? prior[prior.length - 1] : null;

  const knownDevices = new Set(prior.map((h) => h.device_id));
  const knownIps = new Set(prior.map((h) => h.ip));
  const knownAsns = new Set(prior.map((h) => h.asn));

  const within24h = prior.filter((h) => now - Date.parse(h.ts) <= DAY_MS);
  const failedUser1h = prior.filter(
    (h) => h.outcome === "fail" && now - Date.parse(h.ts) <= HOUR_MS,
  ).length;
  const loginsUser1h = prior.filter((h) => now - Date.parse(h.ts) <= HOUR_MS).length;

  const countries24h = new Set(within24h.map((h) => h.country));
  countries24h.add(event.country);
  const devices24h = new Set(within24h.map((h) => h.device_id));
  devices24h.add(event.device_id);
  const ips24h = new Set(within24h.map((h) => h.ip));
  ips24h.add(event.ip);

  let timeSinceLast: number;
  let velocity: number;
  let countryChanged: number;
  if (!last) {
    timeSinceLast = spec.clip_bounds["time_since_last_login_min"];
    velocity = 0;
    countryChanged = 0;
  } else {
    const dtMin = Math.max(0, (now - Date.parse(last.ts)) / 60000);
    timeSinceLast = dtMin;
    const dist = haversineKm(last.lat, last.lon, event.lat, event.lon);
    const dtH = Math.max(dtMin / 60, 1 / 60); // floor at 1 minute
    velocity = dist / dtH;
    countryChanged = last.country !== event.country ? 1 : 0;
  }

  const distHome = haversineKm(event.home_lat, event.home_lon, event.lat, event.lon);

  const named: Record<string, number> = {
    hour,
    dow,
    is_off_hours: isOffHours(hour, event.active_start, event.active_end),
    account_age_days: event.account_age_days,
    time_since_last_login_min: clip(spec, "time_since_last_login_min", timeSinceLast),
    geo_velocity_kmh: clip(spec, "geo_velocity_kmh", velocity),
    distance_from_home_km: clip(spec, "distance_from_home_km", distHome),
    country_is_home: event.country === event.home_country ? 1 : 0,
    country_changed_from_last: countryChanged,
    device_is_known: knownDevices.has(event.device_id) ? 1 : 0,
    ip_is_known: knownIps.has(event.ip) ? 1 : 0,
    asn_is_known: knownAsns.has(event.asn) ? 1 : 0,
    failed_attempts_user_1h: clip(spec, "failed_attempts_user_1h", failedUser1h),
    failed_attempts_ip_1h: clip(spec, "failed_attempts_ip_1h", ipFailLast1h),
    distinct_countries_24h: clip(spec, "distinct_countries_24h", countries24h.size),
    distinct_devices_24h: clip(spec, "distinct_devices_24h", devices24h.size),
    distinct_ips_24h: clip(spec, "distinct_ips_24h", ips24h.size),
    logins_user_1h: clip(spec, "logins_user_1h", loginsUser1h),
    device_type_code: encodeCategorical(spec, "device_type", event.device_type),
    os_code: encodeCategorical(spec, "os", event.os),
    auth_method_code: encodeCategorical(spec, "auth_method", event.auth_method),
  };

  const vector = spec.feature_order.map((f) => named[f]);
  return { vector, named };
}

File: ml/sentinel_ml/features.py
"""Causal feature engineering for Sentinel.

For each event we derive model inputs using **only information available before
that event** — the exact constraint the online scorer faces. The reference
implementation here streams through the time-sorted log once, maintaining the
same per-user and per-IP state the TypeScript scorer reconstructs from recent
rows in Postgres.

Feature windowing conventions (kept identical on both sides):

  * ``failed_attempts_*`` and ``logins_user_1h`` count *prior* events only — the
    current attempt's outcome is not known at decision time.
  * ``distinct_*_24h`` count distinct values across prior events in the window
    **plus** the current event (it reflects state as of "now").
  * ``*_is_known`` flags compare against everything seen for the user strictly
    before this event.
"""

from __future__ import annotations

from collections import defaultdict, deque

import numpy as np
import pandas as pd

from .config import CLIP_BOUNDS, FEATURE_ORDER, encode_categorical
from .geo import haversine_km

HOUR = pd.Timedelta(hours=1)
DAY = pd.Timedelta(hours=24)


def _is_off_hours(hour: int, start: int, end: int) -> int:
    end_mod = end % 24
    if start == end_mod:
        return 0
    if start < end_mod:
        active = start <= hour < end_mod
    else:  # window wraps past midnight
        active = hour >= start or hour < end_mod
    return 0 if active else 1


def _clip(name: str, value: float) -> float:
    bound = CLIP_BOUNDS.get(name)
    if bound is None:
        return float(value)
    return float(min(value, bound))


def compute_features(events: pd.DataFrame) -> pd.DataFrame:
    """Return one feature row per event, in time order.

    Output columns: identifiers + ``FEATURE_ORDER`` + ``is_ato`` + a few raw
    fields kept for the dashboard / analysis.
    """
    events = events.sort_values("ts").reset_index(drop=True)

    # per-user state
    last_ts: dict[str, pd.Timestamp] = {}
    last_loc: dict[str, tuple[float, float]] = {}
    last_country: dict[str, str] = {}
    known_devices: dict[str, set[str]] = defaultdict(set)
    known_ips: dict[str, set[str]] = defaultdict(set)
    known_asns: dict[str, set[int]] = defaultdict(set)
    user_recent: dict[str, deque] = defaultdict(deque)  # (ts, country, device, ip, outcome)
    # per-ip state (across users) for stuffing detection
    ip_recent: dict[str, deque] = defaultdict(deque)  # (ts, outcome)

    out_rows: list[dict] = []

    for row in events.itertuples(index=False):
        uid = row.user_id
        ts = row.ts
        hour = int(ts.hour)
        dow = int(ts.dayofweek)

        # ---- evict stale state from sliding windows ----
        urec = user_recent[uid]
        while urec and (ts - urec[0][0]) > DAY:
            urec.popleft()
        irec = ip_recent[row.ip]
        while irec and (ts - irec[0][0]) > HOUR:
            irec.popleft()

        # ---- history-based features (prior events only) ----
        prev_ts = last_ts.get(uid)
        if prev_ts is None:
            tsl = CLIP_BOUNDS["time_since_last_login_min"]
            velocity = 0.0
            country_changed = 0
        else:
            dt_min = max(0.0, (ts - prev_ts).total_seconds() / 60.0)
            tsl = dt_min
            plat, plon = last_loc[uid]
            dist = haversine_km(plat, plon, float(row.lat), float(row.lon))
            dt_h = max(dt_min / 60.0, 1.0 / 60.0)  # floor at 1 minute to bound velocity
            velocity = dist / dt_h
            country_changed = int(last_country[uid] != row.country)

        failed_user_1h = sum(
            1 for (t, _c, _d, _i, o) in urec if o == "fail" and (ts - t) <= HOUR
        )
        logins_user_1h = sum(1 for (t, *_rest) in urec if (ts - t) <= HOUR)
        failed_ip_1h = sum(1 for (t, o) in irec if o == "fail")

        # distinct over prior-in-window + current
        countries_24h = {c for (_t, c, _d, _i, _o) in urec} | {row.country}
        devices_24h = {d for (_t, _c, d, _i, _o) in urec} | {row.device_id}
        ips_24h = {i for (_t, _c, _d, i, _o) in urec} | {row.ip}

        dist_home = haversine_km(
            float(row.home_lat), float(row.home_lon), float(row.lat), float(row.lon)
        )

        feat = {
            "hour": hour,
            "dow": dow,
            "is_off_hours": _is_off_hours(hour, int(row.active_start), int(row.active_end)),
            "account_age_days": float(row.account_age_days),
            "time_since_last_login_min": _clip("time_since_last_login_min", tsl),
            "geo_velocity_kmh": _clip("geo_velocity_kmh", velocity),
            "distance_from_home_km": _clip("distance_from_home_km", dist_home),
            "country_is_home": int(row.country == row.home_country),
            "country_changed_from_last": country_changed,
            "device_is_known": int(row.device_id in known_devices[uid]),
            "ip_is_known": int(row.ip in known_ips[uid]),
            "asn_is_known": int(int(row.asn) in known_asns[uid]),
            "failed_attempts_user_1h": _clip("failed_attempts_user_1h", failed_user_1h),
            "failed_attempts_ip_1h": _clip("failed_attempts_ip_1h", failed_ip_1h),
            "distinct_countries_24h": _clip("distinct_countries_24h", len(countries_24h)),
            "distinct_devices_24h": _clip("distinct_devices_24h", len(devices_24h)),
            "distinct_ips_24h": _clip("distinct_ips_24h", len(ips_24h)),
            "logins_user_1h": _clip("logins_user_1h", logins_user_1h),
            "device_type_code": encode_categorical("device_type", row.device_type),
            "os_code": encode_categorical("os", row.os),
            "auth_method_code": encode_categorical("auth_method", row.auth_method),
        }

        out_rows.append(
            {
                "event_id": row.event_id,
                "user_id": uid,
                "ts": ts,
                "country": row.country,
                "outcome": row.outcome,
                "attack_type": row.attack_type,
                "is_ato": int(row.is_ato),
                **feat,
            }
        )

        # ---- commit current event to state (now it's "known") ----
        known_devices[uid].add(row.device_id)
        known_ips[uid].add(row.ip)
        known_asns[uid].add(int(row.asn))
        urec.append((ts, row.country, row.device_id, row.ip, row.outcome))
        irec.append((ts, row.outcome))
        last_ts[uid] = ts
        last_loc[uid] = (float(row.lat), float(row.lon))
        last_country[uid] = row.country

    df = pd.DataFrame(out_rows)
    # guarantee column presence + order for the model matrix
    for col in FEATURE_ORDER:
        assert col in df.columns, f"missing feature column {col}"
    return df


def feature_matrix(features: pd.DataFrame) -> np.ndarray:
    """Extract the ordered float32 model-input matrix."""
    return features[FEATURE_ORDER].to_numpy(dtype=np.float32)


if __name__ == "__main__":
    from .config import DATA_PATH, FEATURES_PATH

    ev = pd.read_parquet(DATA_PATH)
    feats = compute_features(ev)
    feats.to_parquet(FEATURES_PATH)
    print(f"Computed features for {len(feats):,} events -> {FEATURES_PATH}")
    print(feats[FEATURE_ORDER].describe().T.to_string())
